Add a C version of the VM event hooks

fsfod opened the pull request:
Add a C version of the VM event hooks that the jitlog will be built on top of.

ltratt commented:
I think this is one for Javier. I've invited him to the repo; when he accepts I'll add him as a reviewer.

ltratt commented:
Hello @javierguerragiraldez :) It would be great to get a review from you on this.

ltratt commented:
@javierguerragiraldez Tom tells me he's done adding commits to this PR, so it's now ready for your review.

@fsfod In order that Javier can do a review, I think you need to give him some more information on the PR about a) motivation for this branch b) how to use it (roughly speaking). 

Thanks both!

fsfod commented:
Its the building blocks that the JIT log will be built on top of. Based on building stuff like this before in Lua I ran into limitations on what data i wanted to access and performance issues because the Lua based event system hooks can't be JIT'ed.
I expanded most the vmevent call macros to also call the my native callback and added new events that are only sent to the native callback. I just hard coded the VM event printer to be always on for testing.

javierguerragiraldez commented on src/lj_state.c:
```diff
@@ -180,6 +181,8 @@ static void close_state(lua_State *L)
     g->allocf(g->allocd, G2GG(g), sizeof(GG_State), 0);
 }
 
+LUA_API void vmevent_printer_start(lua_State *L);
```
conflicts with declaration in `lj_vmevent_printer.h`

fsfod replied:
i just added that temporally to force the event printer to always run when a state is created to make testing simpler

javierguerragiraldez replied:
this conflict persists.  the return type is different between this and the one in `lj_vmevent_printer.h`

javierguerragiraldez commented on src/luajit.h:
```diff
@@ -73,6 +73,13 @@ LUA_API void luaJIT_profile_stop(lua_State *L);
+typedef void(*luaJIT_vmevent_callback)(void *data, lua_State *L,
+                                       int eventid, void *eventdata);
+
+LUA_API int luaJIT_vmevent_sethook(lua_State *L, luaJIT_vmevent_callback cb, void *data);
```
am i right that this `*L` must be the main state (as opposed as a thread/coroutine)?  what about the `*L` that the callback gets, is that also the main one? or would be the current coroutine triggering the event? (yes, i'm really interested in coroutine context)

fsfod replied:
L is really only used as a means to get at the global and jit states. I could add an event for coroutine switching that the jitlog could emit a marker for.

ltratt commented:
Hello @javierguerragiraldez -- do you have any other comments or do you think this is ready for possible merging?

javierguerragiraldez commented:
I think the changes to the JIT (the new events) are very good, simple and readable.  it's the example callback the one i don't like.
i think i wrote somewhere that it's too complex to be just an example, but almost any alternative would have to reuse parts of it.

ltratt commented:
OK. @fsfod do you think you can address Javier's comment?

fsfod commented:
@javierguerragiraldez Most of the callbacks/event handlers were copied from the wip JIT log code.  They initially started out as something close as to what the vmprinter does but i recently started use the skeleton of vmprinter as the base for the jitlog since i cleaned up there interface and added a proper non static context

ltratt commented:
@fsfod Understood, but just because previous things were a bit more complex than necessary doesn't mean we have to do the same :) Do you think you can clean things up a bit?

fsfod commented:
Really the only two things to make it is simpler are removing event filtering and switch to just printing out event name and nothing else

javierguerragiraldez commented:
ok, i'll do a more detailed review of the vmprinter.  just to be clear, what's the specific intent?  as a foundation to be used as is?  or extended by an embedder? an example to be replaced?

fsfod commented:
First to sanity check the events are generated and second as debug layer jitlog. i was thinking maybe also creating some units based on its output.  As i was working on it i thought it be useful as standalone tool where you can't use the lua based jitdump modules that's why i added the user settable output stream.

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+#include "lauxlib.h"
+
+typedef struct VMPrintContext {
+  VMEventContext user;
```
these two types, `VMPrintContext` and `VMEventContext` are confusing.  from the code, i see that the first one is the private data for this callback module, and the second is an extra facility to set the output file.   but from the names i'd believe that `VMEventContext` is a more generic structure (for all 'event related' stuff)

fsfod replied:
fixed by renaming by VMEventContext to VMPrintUserContext

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+  GCtrace *T = (GCtrace*)eventdata;
+  BCPos pc = J->pt ? (int32_t)proto_bcpos(J->pt, J->pc) : -1;
+  GCproto *pt = J->pt;
+  lua_assert(J->pt);
```
is C89 compatibility a requirement?  some of these would be easier to read if closer to where they're used.   also not sure if it actually helps to use those instead of the `J` fields.

fsfod replied:
I had been stuck with having to use c89 compilers in the past for many years while working on LuaJIT so thats just how i normally code for LuaJIT. I assume Mike Paul writes for the some lowest common denominator in compilers as well but i don't think it matters for this file.

ltratt replied:
I don't think we need to be hamstrung by C89: we should assume C11. If a platform doesn't support that, we've got much bigger worries...

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+    fprintf(context->user.output, "START(%d): %s %s %d\n", T->traceno, start,  proto_chunknamestr(pt), lj_debug_line(pt, pc));
+  } else {
+    fprintf(context->user.output, "START(%d): side %s %d\n", T->traceno, proto_chunknamestr(pt), lj_debug_line(pt, pc));
+  }
```
please factorize the commonality of these o these options.  i'm sure it would also help with the variables use.

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+    pc = context->lastpc;
+  } else {
+    pc = proto_bcpos(pt, J->pc);
+  }
```
for readability:
```
if (J->pt) {
    pt = ....
    pc = ...
} else {
    pt = ....
    pc = ...
}
```

fsfod replied:
Fixed by replacing with a shared function for getting the last lua location

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+    fprintf(context->user.output, "STOP(%d): Compiled exit to %s %d\n", T->traceno, proto_chunknamestr(pt), getlastline(context, pt, pc));
+  } else {
+    fprintf(context->user.output, "STOP(%d): %s %d\n", T->traceno, proto_chunknamestr(pt), getlastline(context, pt, pc));
+  }
```
same as on `vmevent_tracestart`. here, it could be like:
```
fprintf(context->user.output, "STOP(%d): %s %s %d\n",
      T->traceno,
      J->cur.linktype == LJ_TRLINK_INTERP ? "Compiled exit to" : "",
      proto_chunknamestr(pt),
      getlastline(context, pt, pc));
```

fsfod replied:
Fixed by simplifying to a single call and skipping printing redundant stuff the trace start already prints

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+    context->lastlua = J->fn;
+  }
+
+  if (pc != -1) {
```
this condition is tied to the same condition on the `pc` definition, and is the only use of that variable.  i think simply removing both variables would make the whole function much more readable.

fsfod replied:
Fixed by simplifying the whole function

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+
+static void vmevent_exit(VMPrintContext *context, jit_State *J, void *exitState)
+{
+  int gcexit = J2G(J)->gc.state == GCSatomic || J2G(J)->gc.state == GCSfinalize;
```
again, a variable with a single use and non-obvious initialization only helps to pull apart related parts of the code.

fsfod replied:
I had separated it out as a variable because the meaning was not obvious and it would of made the fprintf less readable if it was written inline. I renamed the variable and added a comment

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+static void vmevent_protobl(VMPrintContext *context, jit_State *J, void *protopc)
+{
+  GCproto *pt = ((void**)protopc)[0];
+  BCIns *pc = ((void**)protopc)[1];
```
what happened here? aren't structs available?  i think all these second-level handlers should use typed arguments and the cast happen in the main callback.

fsfod replied:
fixed replaced with a proper struct. i think i was just trying to get it working quickly and forgot about it.

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,262 @@
+
+LUA_API void vmevent_printer_shutdown(lua_State *L);
+
+void vmevent_protoloaded(VMPrintContext *context, jit_State *J, void *eventdata)
```
should this be static?

fsfod replied:
yes i missed that one

ltratt commented:
@fsfod Is this ready for rereview or ...?

fsfod commented:
maybe one more or two more things to finish first

ltratt commented:
OK -- let us know when you're ready!

fsfod commented:
@ltratt it should be ready now

ltratt commented:
Thanks! @javierguerragiraldez do you want to have a look at the updates?

javierguerragiraldez commented on src/lj_vmevent_printer.h:
```diff
@@ -0,0 +1,24 @@
+  EVENT_CLASS_PROTO_LOADED = 0x8,
+} VM_EVENT_CLASS;
+
+typedef void (*luaJIT_vmprint_printer)(struct VMPrintUserContext *ctx, const char *msg, uint32_t msglen);
```
undefined use of struct `VMPrintUserContext`, the warning can be silenced by adding a partial struct declaration.  just a line like
```
struct VMPrintUserContext;
```
above this.

fsfod replied:
fixed

javierguerragiraldez replied:
almost!  now i'm getting warnings about typedef redeclaration.  the quietest that works for me is to put the `typedef` in the initial partial struct.  The latter complete is just a struct definitnion.
```
typedef struct VMPrintUserContext VMPrintUserContext;
typedef void (*luaJIT_vmprint_printer)(VMPrintUserContext *ctx, const char *msg, uint32_t msglen);

struct VMPrintUserContext {
  void *userdata;                /* Optional pointer that the user can set to arbitrary data to keep track of the context */
  luaJIT_vmprint_printer output; /* User overridable function to output the messages generated */
  VM_EVENT_CLASS filter;         /* Bit mask of events to filter out of being printed */
};
```

fsfod replied:
@javierguerragiraldez  what compiler and flags are you using to get those warnings?

javierguerragiraldez replied:
actually it was flagged by the IDE's syntax checker.  gcc shows it with `-Wpedantic`.

yes, with lots of noise, but this is the only warning directly attributable to these files

fsfod replied:
fixed

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+
+static void vmevent_tracestart(VMPrintContext *context, jit_State *J, void *eventdata)
+{
+  GCtrace *T = (GCtrace*)eventdata;
```
please use typed pointers.

fsfod replied:
fixed

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+  vmevent_log(context, "STOP(%d): %s:%d, link '%s'\n", T->traceno, proto_chunknamestr(pt), lj_debug_line(pt, pc), getlinktype(T->linktype));
+}
+
+static void vmevent_tracebc(VMPrintContext *context, jit_State *J, void *eventdata)
```
unused, untyped `eventdata` argument.  might be better left off.

fsfod replied:
removed

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+  }
+}
+
+static void vmevent_exit(VMPrintContext *context, jit_State *J, void *exitState)
```
unused, untyped `exitState` argument.

fsfod replied:
Added a proper typed struct event data but we can't really use it much here

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+  vmevent_log(context, "ABORT(%d): reason %d, %s:%d\n", T->traceno, abortreason, proto_chunknamestr(pt), lj_debug_line(pt, pc));
+}
+
+static void vmevent_traceflush(VMPrintContext *context, jit_State *J, uintptr_t data)
```
useless `data` argument

fsfod replied:
I was going to use it as like a reason enum but worried about the number of places in existing code i would have to modify to to feed through the reason

javierguerragiraldez replied:
could (some of) that info be pulled from `J` instead?

ltratt replied:
Sounds like we should just get rid of it and maybe add it back in another PR if needed?

fsfod replied:
trying to use info from J is just guessing. Some flush trigger reasons are not from hitting limits and trace ids can be recycled if a proto is garbage collected that has root traces

javierguerragiraldez replied:
:-(

ltratt replied:
So I'm a bit confused: if we need this parameter to differentiate things, shouldn't we be using it? Or is it too much for this PR? Or ...?

fsfod replied:
well i added it but the pr ever grows

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+  if (context->user.filter & EVENT_CLASS_GC) {
+    return;
+  }
+  vmevent_log(context, "GCSTATE: %s, totalmem %llukb, strings %d\n", getgcsname(newstate), (long long int)(J2G(J)->gc.total/1024), J2G(J)->strnum);
```
line too long.

fsfod replied:
fixed

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+  vmevent_log(context, "GCSTATE: %s, totalmem %llukb, strings %d\n", getgcsname(newstate), (long long int)(J2G(J)->gc.total/1024), J2G(J)->strnum);
+}
+
+LUA_API void vmevent_printer_shutdown(lua_State *L);
```
what's this?

fsfod replied:
fixed removed

javierguerragiraldez commented on src/lj_jit.h:
```diff
@@ -389,6 +389,9 @@ typedef struct FoldState {
 
+typedef void(*luaJIT_vmevent_callback)(void *data, lua_State *L,
+                                       int eventid, void *eventdata);
+
```
conflicts with definition in `luajit.h`

fsfod replied:
I can't really be including the luajit.h header here I will have to make a new header for it

fsfod replied:
fixed with a new header that the eventid enum is also moved to as well as eventdata structs

javierguerragiraldez commented on src/lj_vmevent_printer.c:
```diff
@@ -0,0 +1,323 @@
+
+static void default_ouput(VMPrintUserContext* ctx, const char *msg, uint32_t msglen)
+{
+  printf("%s", msg);
```
could this be conditionally compiled?  i'd like to set something in the makefile or `luaconf.h` and avoid linking stdio in the JIT core.

fsfod replied:
Luajit already uses stdio its the only way the lua based vmevent system can signal an uncaught Lua error happened in a vmevent hook.

fsfod commented:
@javierguerragiraldez i think fixed the remaining issues with eventdata and there arguments

javierguerragiraldez commented:
the code now feels really complete and very readable; i think the only worrying point is the `vmevent_printer_start()` redeclaration in `lj_state.c` vs. the one in `lj_vmevent_printer.h`.

fsfod commented:
that line will be reverted before i merge this it only there  to allow testing without it you would have to make your own little test program to enable the printer

fsfod commented:
@ltratt @javierguerragiraldez  i fixed the last typedef issue can squash?

ltratt commented:
Two minor comments: the commit "TEMP automatically enable vmevent printer on state creation " seems weird. Should that be squashed into something else?

The commit "Add VMEVENT_TRACE_FLUSH thats triggered when a full trace happends" has two typos. "thats" -> "that's". "happends" -> "happens".

fsfod commented:
it will be removed once you two are ok with the last 2 new commits and commit messages. I found some issues when vmevents are disabled at compile time with LUAJIT_DISABLE_VMEVENT and the last 2 commits fix it

ltratt commented:
If @javierguerragiraldez is OK with those two commits, then please go ahead and reword and squash (but only when Javier says OK to do so!).

fsfod commented:
@javierguerragiraldez did you get a chance to look yet it?

javierguerragiraldez commented:
yes (finally).  as a whole, i like it very much.  the only weird thing is that 'TEMP' commit.

fsfod commented:
@javierguerragiraldez I just did the final rebase the temp commit should now be gone

ltratt commented:
@fsfod if you rebase this against master, then i think i can merge this

fsfod commented:
@ltratt its now based off master but the change should be a nop except for what branch we merge in to

ltratt commented:
Agreed, I think we can merge unless you say otherwise.

fsfod commented:
go for it
